posts/inside-rust/2022-08-11-diagnostic-effort.md

```diff
@@ -0,0 +1,336 @@
+   then...
+    - [...use a diagnostic derive](#Using-a-diagnostic-derive…).
+2. Otherwise...
+    - [...manually implement `SessionDiagnostic`](#Manually-implementing-SessionDiagnostic…).
+
+In both cases, diagnostics are represented as types. Representing diagnostics
+using types is a goal of the diagnostic working group as it helps separate
+diagnostic logic from the main code paths.
+
+Every diagnostic type should implement `SessionDiagnostic` (either manually or
+automatically). In the `SessionDiagnostic` trait, there's a member function
+which converts the trait into a `Diagnostic` to be emitted.
+
+#### Using a diagnostic derive...
+Diagnostic derives (either `SessionDiagnostic` for whole diagnostics,
+`SessionSubdiagnostic` for parts of diagnostics or `DecorateLint` for lints)
+can be used to automatically implement a diagnostic trait.
+
+To start, create a new type in the `errors` module of the current crate (e.g.
+`rustc_typeck::errors` or `rustc_borrowck::errors`) named after your
+diagnostic. In our example, that might look like:
+
+```rust=
+struct ReturnTypeArrow {
+
+}
+```
+
+Next, we'll need to add fields with all the information we need - that's just a
+`Span` for us:
+
+```rust=
+struct ReturnTypeArrow {
+    span: Span,
+}
+```
+
+In most cases, this will just be the `Span`s that are used by the original
+diagnostic emission logic and values that are interpolated into diagnostic
+messages.
+
+After that, we should add the derive, add our error attribute and annotate the
+primary `Span` (that was given to `struct_span_err`).
+
+```rust=
+#[derive(SessionDiagnostic)]
+#[error(parser::return_type_arrow)]
+struct ReturnTypeArrow {
+    #[primary_span]
+    span: Span,
+}
+```
+
+Each diagnostic should have a unique slug. By convention, these always start
+with the crate that the error is related to (`parser` in this example). This
+slug will be used to find the actual diagnostic message in our translation
+resources, which we'll see shortly.
+
+Finally, we need to add any labels, notes, helps or suggestions:
+
+```rust=
+#[derive(SessionDiagnostic)]
+#[error(parser::return_type_arrow)]
+struct ReturnTypeArrow {
+    #[primary_span]
+    #[suggestion(applicability = "machine-applicable", code = "->")]
+    span: Span,
+}
+```
+
+In this example, there's just a single suggestion - to replace the `:` with
+a`->`.
+
+Before we're finished, we have to [add the diagnostic messages to the
+translation resources..](#Adding-translation-resources…)
+
+For more documentation on diagnostic derives, see the [diagnostic structs
+chapter of the `rustc` dev guide][diag_struct].
+
+#### Manually implementing `SessionDiagnostic`...
+Some diagnostics are too complicated to be generated from a diagnostic type
+using the diagnostic derive. In these cases, `SessionDiagnostic` can be
+implemented manually.
+
+Using the same type as in ["Using a diagnostic
+derive..."](#Using-a-diagnostic-derive…), we can implement `SessionDiagnostic`
```

Comment:
```suggestion
derive..."](#Using-a-diagnostic-derive...), we can implement `SessionDiagnostic`
```

Reply:
Good catch. I should actually check these references work, I just copied them from the draft hackmd where they definitely did.

Reply:
Alright, fixed these.